Show team versus team scores and winner at the results screen

peppy opened the pull request:
- [x] Depends on #14179.

https://user-images.githubusercontent.com/191335/128983051-71abf307-3383-4c92-9b87-5348da5790b7.mp4

I'm sure we'll come up with a better design for this in the future, but it should do for now.

peppy committed:
Fix match score display not doing an initial value update

peppy committed:
Convey and show team scores at the multiplayer results screen

peppy committed:
Add winner text

peppy committed:
Move text to avoid overlapping the results panel

bdach commented:
I've gotta say, even for a temporary design, what is depicted on the video in the OP seems quite raw - by which I specifically mean both the "team wins" text and the score display text blending in with the background contents, and the score display overlapping with the score panels...

How about something like this, perhaps?

https://user-images.githubusercontent.com/20418176/129111923-0ea9b345-947e-44db-8874-89e4b91d5693.mp4

<details>
<summary>Patch for the above</summary>

```diff
diff --git a/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerResultsScreen.cs b/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerResultsScreen.cs
index 0e4655e336..6d9dbde6f5 100644
--- a/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerResultsScreen.cs
+++ b/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerResultsScreen.cs
@@ -6,8 +6,10 @@
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
+using osu.Framework.Graphics.Colour;
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Effects;
+using osu.Framework.Graphics.Shapes;
 using osu.Game.Graphics;
 using osu.Game.Graphics.Sprites;
 using osu.Game.Localisation;
@@ -15,6 +17,7 @@
 using osu.Game.Scoring;
 using osu.Game.Screens.OnlinePlay.Playlists;
 using osu.Game.Screens.Play.HUD;
+using osuTK;
 
 namespace osu.Game.Screens.OnlinePlay.Multiplayer
 {
@@ -22,6 +25,7 @@ public class MultiplayerResultsScreen : PlaylistsResultsScreen
     {
         private readonly SortedDictionary<int, BindableInt> teamScores;
 
+        private Container winnerBackground;
         private Drawable winnerText;
 
         public MultiplayerResultsScreen(ScoreInfo score, long roomId, PlaylistItem playlistItem, SortedDictionary<int, BindableInt> teamScores)
@@ -36,6 +40,10 @@ public MultiplayerResultsScreen(ScoreInfo score, long roomId, PlaylistItem playl
         [BackgroundDependencyLoader]
         private void load()
         {
+            ScorePanelList.Anchor = ScorePanelList.Origin = Anchor.TopCentre;
+            ScorePanelList.Scale = new Vector2(0.9f);
+            ScorePanelList.Y = 75;
+
             if (teamScores.Count == 2)
             {
                 var redScore = teamScores.First().Value;
@@ -55,6 +63,32 @@ private void load()
                         Team1Score = { BindTarget = redScore },
                         Team2Score = { BindTarget = blueScore },
                     },
+                    winnerBackground = new Container
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Alpha = 0,
+                        Children = new[]
+                        {
+                            new Box
+                            {
+                                RelativeSizeAxes = Axes.X,
+                                Height = 250,
+                                Anchor = Anchor.Centre,
+                                Origin = Anchor.BottomCentre,
+                                Colour = ColourInfo.GradientVertical(Colour4.Black.Opacity(0), Colour4.Black.Opacity(0.4f))
+                            },
+                            new Box
+                            {
+                                RelativeSizeAxes = Axes.X,
+                                Height = 250,
+                                Anchor = Anchor.Centre,
+                                Origin = Anchor.TopCentre,
+                                Colour = ColourInfo.GradientVertical(Colour4.Black.Opacity(0.4f), Colour4.Black.Opacity(0))
+                            }
+                        }
+                    },
                     (winnerText = new OsuSpriteText
                     {
                         Alpha = 0,
@@ -80,6 +114,7 @@ protected override void LoadComplete()
             using (BeginDelayedSequence(300))
             {
                 winnerText.FadeInFromZero(600, Easing.InQuint);
+                winnerBackground.FadeInFromZero(600, Easing.InQuint);
 
                 winnerText
                     .ScaleTo(10)
@@ -87,6 +122,7 @@ protected override void LoadComplete()
                     .Then()
                     .ScaleTo(1.02f, 1600, Easing.OutQuint)
                     .FadeOut(5000, Easing.InQuad);
+                winnerBackground.Delay(2200).FadeOut(2000);
             }
         }
     }
```

</details>

peppy commented:
sure go ahead and commit that change.

i attempted to do the scaling and layout of the score flow in a proper way but (fixing the base classes) but gave up due to the complexity. i’m okay with the temporary jacket way proposed though, especially since it’s probably going to change in design anyway.

bdach committed:
Scale score panel to remove overlap with team score display

bdach committed:
Add winner text background to increase contrast

bdach committed:
Ensure tests wait for screen load

bdach committed:
Split off multiplayer team results screen to separate class

The previous version tried to keep both normal multiplayer and team
multiplayer results as one screen, but didn't check that team-specific
components aren't null in `LoadComplete()`.

To decrease number of conditional, split off the team results screen to
a separate implementation, and choose one or the other at push time in
`MultiplayerPlayer`, depending on team count.

bdach committed:
Add test coverage for both teams winning

bdach committed:
Handle ties in team vs. results screen

bdach committed:
Merge branch 'master' into multiplayer-team-vs-results

bdach commented:
So I've pushed that above change in, but while I was testing I was noticing stuff that needed fixing, so I kinda ended up taking over this PR a smidge. Apologies if overstepping.

In short, two major changes aside from the above patch:

* The tests weren't waiting for the screen to load, obscuring a failure in the "no teams" case (`LoadComplete()` was always applying transforms to drawables that were only initialised if there were two teams). Initially I started adding checks but the end result looked ugly, so I split off the entire team result screen implementation to a separate class and just made `MultiplayerPlayer` choose the implementation based on team count. I *think* that's cleaner?
* Ties weren't handled properly (blue team would be shown to "win" on the text). I've fixed this, but unsure about the copy used, so feel free to adjust if you don't like it.

smoogipoo commented:
Continuously clicking on the panel causes it to gradually move offscreen: https://drive.google.com/file/d/1W4o6KJ5EYW0qdps57EEgi73SW624jiX_/view?usp=sharing Happens in-game too.

peppy committed:
Merge branch 'master' into multiplayer-team-vs-results

peppy committed:
Fix vertical offset not being handled correctly during score panel detach process

peppy commented:
> Continuously clicking on the panel causes it to gradually move offscreen: https://drive.google.com/file/d/1W4o6KJ5EYW0qdps57EEgi73SW624jiX_/view?usp=sharing Happens in-game too.

Fixed by moving the applications to the parent container. I think this is the simplest way (tried to figure why it's not being applied in the detach process but that thing is... weird).

frenzibyte commented on osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerTeamResults.cs:
```diff
@@ -0,0 +1,96 @@
+            AddUntilStep("wait for loaded", () => screen.IsLoaded);
+        }
+
+        [Test]
+        public void TestDisplayWithoutTeams()
+        {
+            MultiplayerResultsScreen screen = null;
+
+            AddStep("show results screen", () =>
+            {
+                var rulesetInfo = new OsuRuleset().RulesetInfo;
+                var beatmapInfo = CreateBeatmap(rulesetInfo).BeatmapInfo;
+
+                var score = new ScoreInfo
+                {
+                    Rank = ScoreRank.B,
+                    TotalScore = 987654,
+                    Accuracy = 0.8,
+                    MaxCombo = 500,
+                    Combo = 250,
+                    Beatmap = beatmapInfo,
+                    User = new User { Username = "Test user" },
+                    Date = DateTimeOffset.Now,
+                    OnlineScoreID = 12345,
+                    Ruleset = rulesetInfo,
+                };
+
+                PlaylistItem playlistItem = new PlaylistItem
+                {
+                    BeatmapID = beatmapInfo.ID,
+                };
+
+                Stack.Push(screen = new MultiplayerResultsScreen(score, 1, playlistItem));
+            });
+
+            AddUntilStep("wait for loaded", () => screen.IsLoaded);
+        }
```
nit: this case seems to be pointless now that the team results variant have been separated to their own screens.

bdach replied:
The non-team-results variant will have no test coverage if this case is removed. So I'd rather keep it.

frenzibyte replied:
Sure, in a separate test class as it's irrelevant from teams now.

frenzibyte reviewed (dismissed):
Currently multi-spectators do not have this winner message appearing, but supporting that does not seem straightforward given that `MultiSpectatorScreen` is not aware of when a player finishes gameplay.

I think `MultiSpectatorScreen` should transition to a shared results screen instead of having one per player area, allowing for interaction and reusing this `MultiplayerTeamResultsScreen` there. Something for later.

bdach committed:
Split test scene into two to follow class split

bdach committed:
Merge branch 'master' into multiplayer-team-vs-results

bdach commented:
I've noticed an issue with the team score display being potentially off-by-1, but it's not caused by this pull, so I see no reason to hold it up further.
